Update notebooks guide to tutorial format (#1355)
astronomer/docs: learn/execute-notebooks.md

---
title: "Execute a Jupyter notebook with Airflow"

> description: "Run a parameterized Jupyter Notebook using Airflow and the Astro CLI."

Reviewer: ```suggestion
description: "Run a parameterized Jupyter notebook using Airflow and the Astro CLI."
```

- Understand what use cases are ideal for orchestrating Jupyter notebooks with Airflow

## Time to complete

This tutorial takes approximately 30 minutes to complete.

## Assumed knowledge

To get the most out of this guide, you should have an understanding of:

- Airflow operators. See [Operators 101](what-is-an-operator.md).
- Creating Jupyter notebooks. See [Try Jupyter](https://docs.jupyter.org/en/latest/start/index.html). 

## Prerequisites

To complete this tutorial, you need:

- The [Astro CLI](https://docs.astronomer.io/astro/cli/get-started).
- The [Jupyter Notebook](https://jupyter.org/install) package.

## Step 1: Create an Astro project

To run a DAG that executes a Jupyter notebook, you first need to create an Astro project, which contains the set of files necessary to run Airflow locally.

1. Create a new directory for your Astro project:

    ```sh
    mkdir <your-astro-project-name>
    ```

2. Open the directory:

    ```sh
    cd <your-astro-project-name>
    ```

3. Run the following Astro CLI command to initialize an Astro project in the directory:

    ```sh
    astro dev init
    ```

## Step 2: Create a Jupyter notebook

> Create a Jupyter notebook called `example_notebook.ipynb` and save it to the `include/` directory of the Astro project you created in Step 1.

Reviewer: Because this is still just a template and involves the Astro project, I think it would be okay to include it in Step 1. and reduce the step count of this tutorial by 1

## Step 3: Parameterize your Jupyter notebook (optional)

Parameterize any cells in your notebook as needed. If you need to pass any information to your notebook at run time, tag the cell in your notebook as described in the [Papermill usage documentation](https://papermill.readthedocs.io/en/latest/usage-parameterize.html).

>     The following notebook prints a simple statement with the current date. The second cell is parameterized so that the `execution_date` is dynamic.
> 
>     ![Notebook param](/img/guides/parameterized_notebook.png)

Reviewer 2: ```suggestion
The following notebook prints a simple statement with the current date. The second cell is parameterized so that the `execution_date` is dynamic.

![Notebook param](/img/guides/parameterized_notebook.png)
```

I think this should fix the format issue :)

## Step 4: Install supporting packages

> Install the Papermill provider and supporting packages required to run the notebook kernel. The `PapermillOperator` is designed to run a notebook locally, so you need to supply a kernel engine for your Airflow environment to execute the notebook code. For this tutorial we use the `ipykernel` package to run the kernel, but there are other options available such as the `jupyter` package.

Reviewer: Move "The `PapermillOperator` is designed to run a ..... available such as the `jupyter` package" until after the actual `requirements.txt` step, since this information is optional.

> Install the Papermill provider and supporting packages required to run the notebook kernel. The `PapermillOperator` is designed to run a notebook locally, so you need to supply a kernel engine for your Airflow environment to execute the notebook code. For this tutorial we use the `ipykernel` package to run the kernel, but there are other options available such as the `jupyter` package.

Reviewer: ```suggestion
Install the Papermill provider and supporting packages required to run the notebook kernel. The `PapermillOperator` is designed to run a notebook locally, so you need to supply a kernel engine for your Airflow environment to execute the notebook code. This tutorial uses the `ipykernel` package to run the kernel, but there are other options available such as the `jupyter` package.
```

Add the following to the `requirements.txt` file of your Astro project:

>     ```text
>     apache-airflow-providers-papermill
>     ipykernel
>     ```

Reviewer 2: ```suggestion
```text
apache-airflow-providers-papermill
ipykernel
```
```

Reviewer: Agreed with Tamara but watch out when committing this suggestion - GitHub didn't parse the last three ticks correctly and they're missing from the suggestion.

## Step 5: Create your DAG

> Create your DAG with the `PapermillOperator` to execute your notebook. Use your favorite code editor or text editor to copy-paste the following code into a .py file in your project's `dags/` directory:

Reviewer: ```suggestion
Create your DAG with the `PapermillOperator` to execute your notebook. Use your favorite code editor or text editor to copy-paste the following code into a `.py` file in your project's `dags/` directory:
```

```python
from datetime import datetime, timedelta

from airflow import DAG
from airflow.providers.papermill.operators.papermill import PapermillOperator

with DAG(
    dag_id='example_papermill_operator',
    default_args={
        'retries': 0
    },
    schedule_interval='0 0 * * *',
    start_date=datetime(2022, 10, 1),
    template_searchpath='/usr/local/airflow/include',
    catchup=False
) as dag_1:

    notebook_task = PapermillOperator(
        task_id="run_example_notebook",
        input_nb="include/example_notebook.ipynb",
        output_nb="include/out-{{ execution_date }}.ipynb",
        parameters={"execution_date": "{{ execution_date }}"},
    )
```

The `PapermillOperator` requires the following arguments:

- `input_nb`: The notebook you want to run.
- `output_nb`: The path to your output notebook (i.e. the notebook which shows the results of the notebook execution).
- `parameters`: A JSON dictionary of any parameters you are passing to your notebook.

> Note that the built-in `execution_date` Airflow variable is used so that the DAG is idempotent. Parameters for your notebook can come from anywhere, but we highly recommend using Airflow macros and environment variables to avoid hard-coding values in your DAG file.

Reviewer: ```suggestion
Note that the built-in `execution_date` Airflow variable is used so that the DAG is idempotent. Parameters for your notebook can come from anywhere, but Astronomer recommends using Airflow macros and environment variables to avoid hard-coding values in your DAG file.
```

## Step 6: Run your DAG to execute your notebook

> Trigger your DAG to executes the `example_notebook.ipynb` and generate an output notebook named with the execution date. Open the output notebook in your `include/` directory to see the results of the run:

Reviewer 2: ```suggestion
Trigger your DAG to execute the `example_notebook.ipynb` and generate an output notebook with a name that includes the execution date. Open the output notebook in your `include/` directory to see the results of the run:
```

>     ![Output notebook](/img/guides/notebook_output.png)

Reviewer 2: ```suggestion
![Output notebook](/img/guides/notebook_output.png)
```
